osmoutils/slice_helper.go

```diff
@@ -92,3 +93,103 @@ func MergeSlices[T any](slice1, slice2 []T, less LessFunc[T]) []T {
+// Allocates additional space and results in 0(n) space complexity.
+// In place.
+// Parameters:
+// s: slice of type T and size n
+// i: index of the ith smallest element to return where i is in the range [1, n]
+// less: a function that defines the ordering of the elements in the slice.
+// Example:
+// Input: [5, 7, 1, 4, 9] 2, less = func(a, b int) bool { return a < b }
+// Output: 5
+// More information about the algorithm:
+// https://brilliant.org/wiki/median-finding-algorithm/
+// https://www.youtube.com/watch?v=EzeYI7p9MjU
+func iThSmallest[T constraints.Ordered](s []T, i int, less LessFunc[T]) T {
+	if i < 0 || i > len(s) {
+		panic(fmt.Sprintf("i (%d) is out of bounds (%d)", i, len(s)))
+	}
+
+	// Select a pivot by dividing the original slice into subgroups of 5 elements.
+
+	originalLength := len(s)
+
+	// Pre-allocate enough buffer for the medians in case input is large.
+	mediansOfSubSlices := make([]T, 0, originalLength/5+originalLength%5)
+	for i := 0; i < originalLength; i += 5 {
+		// choose either 5 elements, or everything that is remaining
+		// if the last slice ends up being smaller than 5 elements.
+		sliceOfFive := s[i:Min(i+5, originalLength)]
+
+		// sort the subslice of five. Note, that this is sort
+		// is O(1) since only 5 elements.
+		sort.Slice(sliceOfFive, func(i, j int) bool {
+			return less(sliceOfFive[i], sliceOfFive[j])
+		})
+
+		// We append the median of the slice of 5 elements into the median
+		// medians of sub slices.
+		mediansOfSubSlices = append(mediansOfSubSlices, sliceOfFive[len(sliceOfFive)/2])
+	}
+
+	var pivot T
+	numberOfMedians := len(mediansOfSubSlices)
+	if numberOfMedians <= 5 {
+		// Base case of the pivot finding divide and conquer.
+		sort.Slice(mediansOfSubSlices, func(i, j int) bool {
+			return less(mediansOfSubSlices[i], mediansOfSubSlices[j])
+		})
+		pivot = mediansOfSubSlices[numberOfMedians/2]
+	} else {
+		// Use the same algorithm to find the median of the medians of subslices
+		// to pivot on for solving the original problem.
+		pivot = iThSmallest(mediansOfSubSlices, numberOfMedians/2, less)
+	}
+
+	// Note, that partitioning below takes O(n)
+	// By the master's theorem the overall result
+	// gets amortized to O(n) regardless.
+
+	// partition elements
+	smallerThanPivot := make([]T, 0)
+	greaterThanPivot := make([]T, 0)
+	numEqualToPivot := 0
+	for _, cur := range s {
+		if cur == pivot {
+			numEqualToPivot++
+			continue
+		}
+		if less(cur, pivot) {
+			smallerThanPivot = append(smallerThanPivot, cur)
+		} else {
+			greaterThanPivot = append(greaterThanPivot, cur)
+		}
+	}
+	numberOfElementsSmallerThanPivot := len(smallerThanPivot)
+
+	if i < numberOfElementsSmallerThanPivot {
+		return iThSmallest(smallerThanPivot, i, less)
+	} else if i > numberOfElementsSmallerThanPivot {
+		// Handle edge case where there are duplicate pivots and it is the result.
+		if len(greaterThanPivot) == 0 && numEqualToPivot > 1 {
```

Comment:
```suggestion
                // we return pivot, if `i` lays between the sets of elements smaller than pivot and bigger than pivot
		if i < originalLength-len(greaterThanPivot) && i > numberOfElementsSmallerThanPivot && numEqualToPivot > 1 {
```

Reply:
the reason I started digging into this, is because adding:
```
"failing tc": {
			s:           []int{1, 2, 3, 4, 5, 6, 7, 7, 7, 7, 8, 9, 10},
			i:           9,
			less:        less,
			expected:    7,
			expectPanic: false,
		},
```

will fail the test.

Reply:
I think, the current implementation only accounts for a situation, if the pivot is ithSmallest, but only if it is located at the end of the array we are recursing on

Reply:
In the `failing tc` case, we pick 7 as a pivot, but because there are multiple copies of it and the selected pivot 7 is not ithSmallest, we continue recursing on the array of greater elements (which in turn results in out of bounds error, because we left the answer behind)

Reply:
I changed my local branch to the suggested design here, and it successfully passed with the `failing test` test case included

Reply:
also, I removed the `len(greaterThanPivot) == 0` condition, could you please clarify why we need it? maybe I am missing something